# Exploring eBay Car Sales

intro

> question 3

> question 2

### OBJECTIVE

txt

### LIBRARIES
- tpc

### RESOURCES
[DataQuest.io](https://www.dataquest.io)

The dataset was originally scraped and uploaded to Kaggle by the user orgesleka but the dataset as well as the user can't be found anymore. Here's where the dataset can still be found:

[data.world](https://data.world/data-society/used-cars-data)

In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

## Data Cleaning
- Columns rename & understanding
- Assigning proper data types to each column
- Finding wrong date inputs by looking at the outliers
- Dropping irrelevant columns

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The columns are currently following a **camelCase** format, and to ensure we have a better reading as well as align with most other Python projects, we will convert it to **snakecase**.

In [4]:
def camel_to_snake(name):
  name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
  return re.sub('([a-z0-9])([A-Z])', r'\1_\2', name).lower()

In [5]:
new_columns = []

for col in autos.columns:
    new_columns.append(camel_to_snake(col))

autos.columns = new_columns

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_ps   model   odometer  month_of_registration fuel_type  \
0    manuell       158  andere  150,000km                      3       lpg   
1  automatik       286     7er  150,000km                      6    benzin   
2    manuell       102    golf   70,000km                      7    benzin   
3  automatik        71  fortwo   70,000km                      6    benzin   
4    manuell         0   focus  150,000km                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

For some columns it's not clear what do they represent, as well as what is supposed to be the type of data input.

- Should it be a categorical input?
- Should it be a number? If so, with or without decimal places?
- Just simply plain text?
- Dates? ...

To ensure we can properly explore this dataset, let's inspect all columns and possibly drop the ones which doesn't have any added value to our analysis.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox      power_ps  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

By a first look at our dataset we can notice there are some inconsistencies in the columns. As part of our cleaning process we must make sure the following issues are dealt with:

- Inspecting NaN values and find a solution or remove them.
- Assign the proper data-type in each columm.
- Inspecting if all columns are needed for the analysis, and consider dropping them.

For now, the following table represents what we will be focusing on:

| **Column(s)** | **Next Step** |
| :--- | :--- |
| price | Definitely keep it, but make sure data type is a number, not text |
| odometer | Definitely keep it, but make sure data type is a number, not text |
| date_crawled, last_seen, date_created, year_of_registration | Inspect outliers to find possible wrong inputs |
| seller | Inspect why are there only two unique values, and understand what do they represent |
| offer_type | Inspect why are there only two unique values, and understand what do they represent |
| abtest | Inspect why are there only two unique values, and understand what do they represent |

In [7]:
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [8]:
autos['price'] = autos['price'].str.replace("$", "", regex=False)
# Now to replace the commas, is important to consider that some are there to indicate the cents
# and from left to right we will remove the first comma, and replace the second one - if any - with a dot
# allowing to convert it into a float later on.
autos['price'] = autos['price'].str.replace(",", "", 1, regex=False)
autos['price'] = autos['price'].str.replace(",", ".", 1, regex=False)
autos['price'].unique()

array(['5000', '8500', '8990', ..., '385', '22200', '16995'], dtype=object)

In [9]:
autos['price'] = autos['price'].astype(float)

In [10]:
autos['price'].describe()

count     50000.000000
mean       5783.983021
std       11840.063601
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      999999.000000
Name: price, dtype: float64

In [11]:
autos.rename({"price": "price_usd"}, axis=1, inplace=True)
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_usd', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


We managed to convert the 'price' column into a numeric data type, and updated it's column name to include what is the unit the numbers are being displayed. We can already see some inconsistent numbers such as '0.00' and '999999.00', however we will inspect them later.

In [12]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [13]:
autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].astype(float)

In [14]:
autos['odometer'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer, dtype: float64

In [15]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_usd', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


Now the 'odometer' column has also gone through the same changes we did with 'price_usd'. It now has the proper data-types as well as a proper column name.

On the next couple of lines we will focus on understanding the 'price_usd' unrealistic values that were found and mentioned before.

In [16]:
autos['price_usd'].unique().shape

(2354,)

In [17]:
print(autos['price_usd'].value_counts().sort_index(ascending=True).head(50))
print('\n')
print(autos['price_usd'].value_counts().sort_index(ascending=False).head(20))

0.0      1421
1.0       156
2.0         3
3.0         1
5.0         2
8.0         1
9.0         1
10.0        7
11.0        2
12.0        3
13.0        2
14.0        1
15.0        2
17.0        3
18.0        1
20.0        4
25.0        5
29.0        1
30.0        7
35.0        1
40.0        6
45.0        4
47.0        1
49.0        4
50.0       49
55.0        2
59.0        1
60.0        9
65.0        5
66.0        1
70.0       10
75.0        5
79.0        1
80.0       15
89.0        1
90.0        5
99.0       19
100.0     134
110.0       3
111.0       2
115.0       2
117.0       1
120.0      39
122.0       1
125.0       8
129.0       1
130.0      15
135.0       1
139.0       1
140.0       9
Name: price_usd, dtype: int64


999999.0    2
999990.0    1
350000.0    1
345000.0    1
299000.0    1
295000.0    1
265000.0    1
259000.0    1
250000.0    1
220000.0    1
198000.0    1
197000.0    1
194000.0    1
190000.0    1
180000.0    1
175000.0    1
169999.0    1
169000.0    1
163991.0    1
16

In [18]:
print(autos.loc[autos['price_usd'] == 999999.0, ['name', 'price_usd']])

                                   name  price_usd
514    Ford_Focus_Turnier_1.6_16V_Style   999999.0
43049                     2_VW_Busse_T3   999999.0


In [19]:
print(autos.loc[autos['price_usd'] <= 50.0, ['name', 'price_usd']])

                                                    name  price_usd
27                Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE        0.0
55                        Mercedes_E320_AMG_zu_Tauschen!        1.0
64                    Autotransport__Abschlepp_Schlepper       40.0
71     Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...        0.0
80     Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...        0.0
...                                                  ...        ...
49884               Audi_a6_2.5l__Schnaeppchen_nur_heute        0.0
49943                                         Opel_astra        0.0
49960            Ford_KA_zu_verschenken_***Reserviert***        0.0
49974  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...        0.0
49984  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...        0.0

[1688 rows x 2 columns]


In [20]:
autos_clean_price = autos.loc[autos['price_usd'].between(49.0, 351000.0), :]
autos_clean_price.describe(include="all")

date_crawled         name  seller offer_type      price_usd  \
count                 48362        48362   48362      48362   48362.000000   
unique                46697        37275       2          1            NaN   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot            NaN   
freq                      3           76   48361      48362            NaN   
mean                    NaN          NaN     NaN        NaN    5917.823044   
std                     NaN          NaN     NaN        NaN    9081.257003   
min                     NaN          NaN     NaN        NaN      49.000000   
25%                     NaN          NaN     NaN        NaN    1250.000000   
50%                     NaN          NaN     NaN        NaN    3000.000000   
75%                     NaN          NaN     NaN        NaN    7499.000000   
max                     NaN          NaN     NaN        NaN  350000.000000   

       abtest vehicle_type  year_of_registration  gearbox      power_ps  \
count   48362        43884          48362.000000    46103  48362.000000   
unique      2            8                   NaN        2           NaN   
top      test    limousine                   NaN  manuell           NaN   
freq    24903        12572                   NaN    36005           NaN   
mean      NaN          NaN           2004.778793      NaN    117.477834   
std       NaN          NaN             88.823926      NaN    201.004449   
min       NaN          NaN           1000.000000      NaN      0.000000   
25%       NaN          NaN           1999.000000      NaN     73.000000   
50%       NaN          NaN           2004.000000      NaN    107.000000   
75%       NaN          NaN           2008.000000      NaN    150.000000   
max       NaN          NaN           9999.000000      NaN  17700.000000   

        model    odometer_km  month_of_registration fuel_type       brand  \
count   45940   48362.000000           48362.000000     44424       48362   
unique    245            NaN                    NaN         7          40   
top      golf            NaN                    NaN    benzin  volkswagen   
freq     3888            NaN                    NaN     29277       10298   
mean      NaN  125865.762375               5.794508       NaN         NaN   
std       NaN   39630.134489               3.679988       NaN         NaN   
min       NaN    5000.000000               0.000000       NaN         NaN   
25%       NaN  125000.000000               3.000000       NaN         NaN   
50%       NaN  150000.000000               6.000000       NaN         NaN   
75%       NaN  150000.000000               9.000000       NaN         NaN   
max       NaN  150000.000000              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                39403                48362         48362.0  48362.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 34739                 1877             NaN           NaN   
mean                   NaN                  NaN             0.0  50994.674083   
std                    NaN                  NaN             0.0  25744.319018   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30823.000000   
50%                    NaN                  NaN             0.0  49733.000000   
75%                    NaN                  NaN             0.0  71686.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 48362  
unique                38341  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%      

After inspecting the 'price_usd' outliers, it was taken the decision of limiting our dataset to price values which are relevant. Most of the analysis we want to conduct rely on knowing the prices with a certain degree of accuracy, and for that reason we will slice our data to include only:

- Inputs which the price is greater than 49\$
- Inputs which the price is smaller than 351000\$

The inputs found beyond those boundaries represent too much uncertainty. They might've had mispelled characters, poorly filling, human mistake, or even system bugs that generated wrong values.

Now moving ahead to do the same understanding with regards to 'odometer_km':

In [21]:
print(autos_clean_price['odometer_km'].value_counts().sort_index(ascending=False).head(20))

150000.0    31295
125000.0     5045
100000.0     2107
90000.0      1733
80000.0      1414
70000.0      1215
60000.0      1153
50000.0      1013
40000.0       816
30000.0       778
20000.0       759
10000.0       249
5000.0        785
Name: odometer_km, dtype: int64


Differently than the 'price-usd' column, the 'odometer_km' has not shown unrealistic outliers that should be inspected. So no rows will be dropped here.

Moving forward let's evaluate the columns related to dates.

In [22]:
autos_clean_price['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100

2016-03-05    2.535048
2016-03-06    1.406063
2016-03-07    3.602002
2016-03-08    3.329060
2016-03-09    3.306315
2016-03-10    3.221538
2016-03-11    3.256689
2016-03-12    3.699185
2016-03-13    1.565279
2016-03-14    3.666101
2016-03-15    3.428311
2016-03-16    2.948596
2016-03-17    3.157438
2016-03-18    1.290269
2016-03-19    3.475869
2016-03-20    3.777759
2016-03-21    3.730201
2016-03-22    3.298044
2016-03-23    3.225673
2016-03-24    2.942393
2016-03-25    3.149167
2016-03-26    3.227741
2016-03-27    3.107812
2016-03-28    3.488276
2016-03-29    3.413837
2016-03-30    3.370415
2016-03-31    3.184318
2016-04-01    3.372483
2016-04-02    3.552376
2016-04-03    3.856333
2016-04-04    3.655763
2016-04-05    1.304743
2016-04-06    0.316364
2016-04-07    0.138539
Name: date_crawled, dtype: float64

The column 'date_crawled' doesn't seem to have any alarming outlier that could indicate a wrong input.

In [23]:
autos_clean_price['date_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100

2015-06-11    0.002068
2015-08-10    0.002068
2015-09-09    0.002068
2015-11-10    0.002068
2015-12-05    0.002068
                ...   
2016-04-03    3.881146
2016-04-04    3.692982
2016-04-05    1.176544
2016-04-06    0.324635
2016-04-07    0.124064
Name: date_created, Length: 76, dtype: float64

The column 'date_created' doesn't seem to have any alarming outlier that could indicate a wrong input.

In [24]:
autos_clean_price['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001075
2016-03-06    0.004322
2016-03-07    0.005417
2016-03-08    0.007361
2016-03-09    0.009656
2016-03-10    0.010608
2016-03-11    0.012386
2016-03-12    0.023758
2016-03-13    0.008871
2016-03-14    0.012634
2016-03-15    0.015880
2016-03-16    0.016459
2016-03-17    0.028080
2016-03-18    0.007320
2016-03-19    0.015798
2016-03-20    0.020615
2016-03-21    0.020615
2016-03-22    0.021401
2016-03-23    0.018589
2016-03-24    0.019768
2016-03-25    0.019168
2016-03-26    0.016769
2016-03-27    0.015570
2016-03-28    0.020843
2016-03-29    0.022311
2016-03-30    0.024730
2016-03-31    0.023841
2016-04-01    0.022869
2016-04-02    0.024896
2016-04-03    0.025164
2016-04-04    0.024503
2016-04-05    0.124995
2016-04-06    0.221662
2016-04-07    0.132066
Name: last_seen, dtype: float64

The column 'last_seen' doesn't seem to have any alarming outlier that could indicate a wrong input.

In [25]:
autos_clean_price['year_of_registration'].describe()

count    48362.000000
mean      2004.778793
std         88.823926
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

In [26]:
autos_clean_price['year_of_registration'].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2002, 2012, 2011, 2005,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1973, 1967, 1976, 4500, 1987, 1991, 1983, 1960, 1969, 1978,
       1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971, 1966, 1979,
       1981, 1970, 1974, 1975, 5000, 4100, 2019, 1956, 9999, 6200, 1964,
       1959, 1958, 1800, 1948, 1931, 1943, 1941, 1950, 1962, 1927, 1937,
       1929, 1000, 1910, 1957, 1952, 1111, 1955, 1939, 8888, 1954, 1938,
       2800, 5911, 1953, 1951, 4800, 1001, 9000])

As observable above, there are some year values which surely does not correspond to reality. For example we can notice years '1001' in which cars weren't even invented, as well as year '9999' that is in the future. So solve this issue we will create an acceptable year-range and once again slice our dataset.

- **Minimum year**: 1900
- **Maximum year**: 2016

In [27]:
autos_clean_year = autos_clean_price.loc[autos_clean_price['year_of_registration'].between(1900, 2016), :]
autos_clean_year['year_of_registration'].describe()

count    46485.00000
mean      2002.92916
std          7.13266
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: year_of_registration, dtype: float64

In [28]:
autos_clean_year['year_of_registration'].value_counts(normalize=True).head(20)*100

2000    6.711843
2005    6.283747
1999    6.210606
2004    5.806174
2003    5.797569
2006    5.745940
2001    5.664193
2002    5.332903
1998    5.057545
2007    4.891901
2008    4.758524
2009    4.476713
1997    4.173389
2011    3.484995
2010    3.418307
1996    2.934280
2012    2.815962
2016    2.596537
1995    2.596537
2013    1.723136
Name: year_of_registration, dtype: float64

After slicing our dataset and keeping only years with the higher likelihood of being correct, we are left with 46485 inputs. In this last part of our cleaning we will understand more about NaN values and possible columns that could be dropped for not adding any value to our analysis.

In [29]:
autos_clean_year.describe(include="all")

date_crawled                 name  seller offer_type  \
count                 46485                46485   46485      46485   
unique                44944                35624       2          1   
top     2016-04-02 11:37:04  Volkswagen_Golf_1.4  privat    Angebot   
freq                      3                   75   46484      46485   
mean                    NaN                  NaN     NaN        NaN   
std                     NaN                  NaN     NaN        NaN   
min                     NaN                  NaN     NaN        NaN   
25%                     NaN                  NaN     NaN        NaN   
50%                     NaN                  NaN     NaN        NaN   
75%                     NaN                  NaN     NaN        NaN   
max                     NaN                  NaN     NaN        NaN   

            price_usd abtest vehicle_type  year_of_registration  gearbox  \
count    46485.000000  46485        43882           46485.00000    44459   
unique            NaN      2            8                   NaN        2   
top               NaN   test    limousine                   NaN  manuell   
freq              NaN  23953        12572                   NaN    34623   
mean      6006.728099    NaN          NaN            2002.92916      NaN   
std       9199.700841    NaN          NaN               7.13266      NaN   
min         49.000000    NaN          NaN            1910.00000      NaN   
25%       1250.000000    NaN          NaN            1999.00000      NaN   
50%       3149.000000    NaN          NaN            2003.00000      NaN   
75%       7500.000000    NaN          NaN            2008.00000      NaN   
max     350000.000000    NaN          NaN            2016.00000      NaN   

            power_ps  model    odometer_km  month_of_registration fuel_type  \
count   46485.000000  44327   46485.000000           46485.000000     43258   
unique           NaN    244            NaN                    NaN         7   
top              NaN   golf            NaN                    NaN    benzin   
freq             NaN   3695            NaN                    NaN     28453   
mean      118.177735    NaN  125680.004302               5.839475       NaN   
std       185.241766    NaN   39696.081312               3.664450       NaN   
min         0.000000    NaN    5000.000000               0.000000       NaN   
25%        75.000000    NaN  100000.000000               3.000000       NaN   
50%       109.000000    NaN  150000.000000               6.000000       NaN   
75%       150.000000    NaN  150000.000000               9.000000       NaN   
max     17700.000000    NaN  150000.000000              12.000000       NaN   

             brand not_repaired_damage         date_created  nr_of_pictures  \
count        46485               38316                46485         46485.0   
unique          40                   2                   74             NaN   
top     volkswagen                nein  2016-04-03 00:00:00             NaN   
freq          9825               33799                 1812             NaN   
mean           NaN                 NaN                  NaN             0.0   
std            NaN                 NaN                  NaN             0.0   
min            NaN                 NaN                  NaN             0.0   
25%            NaN                 NaN                  NaN             0.0   
50%            NaN                 NaN                  NaN             0.0   
75%            NaN                 NaN                  NaN             0.0   
max            NaN                 NaN                  NaN             0.0   

         postal_code            last_seen  
count   46485.000000                46485  
unique           NaN                37017  
top              NaN  2016-04-07 06:17:27  
freq             NaN                    8  
mean    51117.788900                  NaN  
std     25752.440605                  NaN  
min      1067.000000                  NaN  

In [30]:
autos_clean_year['seller'].value_counts()

privat        46484
gewerblich        1
Name: seller, dtype: int64

In [31]:
autos_clean_year['offer_type'].value_counts()

Angebot    46485
Name: offer_type, dtype: int64

In [32]:
autos_clean_year['abtest'].value_counts()

test       23953
control    22532
Name: abtest, dtype: int64

The column 'seller' and 'offer_type' clearly adds nearly no value, as the vast majority of the data has the same information bound to it.

> **We will drop 'seller' and 'offer_type columns***

In [33]:
final_columns = list(autos_clean_year.columns)
final_columns.pop(final_columns.index('seller'))
final_columns.pop(final_columns.index('offer_type'))

autos_final = autos_clean_year[final_columns]
autos_final.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price_usd   abtest vehicle_type  year_of_registration    gearbox  power_ps  \
0     5000.0  control          bus                  2004    manuell       158   
1     8500.0  control    limousine                  1997  automatik       286   
2     8990.0     test    limousine                  2009    manuell       102   
3     4350.0  control   kleinwagen                  2007  automatik        71   
4     1350.0     test        kombi                  2003    manuell         0   

    model  odometer_km  month_of_registration fuel_type       brand  \
0  andere     150000.0                      3       lpg     peugeot   
1     7er     150000.0                      6    benzin         bmw   
2    golf      70000.0                      7    benzin  volkswagen   
3  fortwo      70000.0                      6    benzin       smart   
4   focus     150000.0                      7    benzin        ford   

  not_repaired_damage         date_created  nr_of_pictures  postal_code  \
0                nein  2016-03-26 00:00:00               0        79588   
1                nein  2016-04-04 00:00:00               0        71034   
2                nein  2016-03-26 00:00:00               0        35394   
3                nein  2016-03-12 00:00:00               0        33729   
4                nein  2016-04-01 00:00:00               0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Brand Analysis
- Include steps of the process

In [50]:
autos_final['brand'].value_counts()

volkswagen        9825
bmw               5111
opel              4988
mercedes_benz     4486
audi              4025
ford              3253
renault           2193
peugeot           1390
fiat              1195
seat               848
skoda              763
nissan             712
mazda              709
smart              659
citroen            654
toyota             593
hyundai            467
sonstige_autos     450
volvo              423
mini               408
mitsubishi         380
honda              365
kia                330
alfa_romeo         309
porsche            281
suzuki             276
chevrolet          265
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru              98
land_rover          98
saab                77
jaguar              73
daewoo              69
trabant             64
rover               62
lancia              49
lada                27
Name: brand, dtype: int64

Looking at the sheer number won't tell us much. So let's look at two specific pieces of information:
- Mean **price** per brand;
- Mean **odometer** value per brand.

And to ensure we get the most insightful analysis, we will consider only brands with more than 500 cars, and group the remaining ones all together.

In [51]:
brand_mean_price = {}
brands = autos_final['brand'].unique()
brand_counts = autos_final['brand'].value_counts().to_dict()
brand_others = []

for brand in brands:
    brand_rows = autos_final['brand'] == brand
    mean_price = autos_final.loc[brand_rows, 'price_usd'].mean()
    if brand_counts[brand] > 500:
        brand_mean_price[brand] = mean_price
    else:
        brand_others.append(brand)

others_bool = autos_final['brand'] == brand_others[0]

for brand in brand_others[1:]:
    others_bool = others_bool | (autos_final['brand'] == brand)

brand_mean_price['others'] = autos_final.loc[others_bool , 'price_usd'].mean()

In [52]:
df_price_means = pd.DataFrame.from_dict(data=brand_mean_price, orient='index', columns=['Mean Price'])
df_price_means.sort_values(by='Mean Price', ascending=False)

Mean Price
audi           9373.771429
mercedes_benz  8683.419750
others         8514.641212
bmw            8375.407663
skoda          6393.018349
volkswagen     5423.855787
toyota         5167.091062
nissan         4750.063202
seat           4423.116745
mazda          4112.596615
citroen        3820.914713
ford           3764.375687
smart          3591.021244
peugeot        3100.692806
opel           2997.959438
fiat           2818.456067
renault        2483.889193

**Takeaways**:

- The difference between the TOP 4 and the next ones to come is pretty considerable;
- We will avoid consider 'Others' as it's grouped as it is exactly due to lack of data (brands with less than 500 cars);
- Audi, Mercedes Benz and BMW are a clear cut at the top;
- Skoda, Volkswagen and Toyota represents a middle ground.

Now let's look at the average odometer value per brand.

In [53]:
brand_mean_odo = {}

for brand in brands:
    brand_rows = autos_final['brand'] == brand
    mean_odo = autos_final.loc[brand_rows, 'odometer_km'].mean()
    if brand_counts[brand] > 500:
        brand_mean_odo[brand] = mean_odo
    else:
        brand_others.append(brand)

brand_mean_odo['others'] = autos_final.loc[others_bool , 'odometer_km'].mean()

In [54]:
df_odo_means = pd.DataFrame.from_dict(data=brand_mean_odo, orient='index', columns=['Mean Odometer'])
df_odo_means.sort_values(by='Mean Odometer', ascending=False)

Mean Odometer
bmw            132655.057719
mercedes_benz  130927.329469
opel           129444.667201
audi           129254.658385
volkswagen     128759.796438
renault        128144.094847
peugeot        127122.302158
mazda          124464.033850
ford           124272.978789
seat           121303.066038
citroen        119525.993884
nissan         118370.786517
fiat           117066.945607
toyota         115944.350759
others         112474.906514
skoda          110884.665793
smart           99590.288316

**Takeaways**:

- BMW, Mercedes Benz and Audi figure in both Price and Odometer top ranking, meaning the aggregated value of the vehicle might not be correlated to the odometer value.
- The variance from TOP 1 to 7 is nearly 5000 km, which is not a very significant value.
- The lasgest cut actually comes from the bottom, in which Smart is more than 10 000 km a part from Skoda.

Now let's look more deeply into the correlation between both Price and Odometer.

In [61]:
df_price_odo = pd.concat([df_price_means, df_odo_means], axis=1)
df_price_odo['km per usd'] = df_price_odo['Mean Odometer'] / df_price_odo['Mean Price']
df_price_odo.sort_values(by='km per usd', ascending=False)

Mean Price  Mean Odometer  km per usd
renault        2483.889193  128144.094847   51.590101
opel           2997.959438  129444.667201   43.177591
fiat           2818.456067  117066.945607   41.535842
peugeot        3100.692806  127122.302158   40.998032
ford           3764.375687  124272.978789   33.012905
citroen        3820.914713  119525.993884   31.282037
mazda          4112.596615  124464.033850   30.264100
smart          3591.021244   99590.288316   27.733138
seat           4423.116745  121303.066038   27.424794
nissan         4750.063202  118370.786517   24.919834
volkswagen     5423.855787  128759.796438   23.739532
toyota         5167.091062  115944.350759   22.438999
skoda          6393.018349  110884.665793   17.344650
bmw            8375.407663  132655.057719   15.838639
mercedes_benz  8683.419750  130927.329469   15.077853
audi           9373.771429  129254.658385   13.788971
others         8514.641212  112474.906514   13.209588

The **'km per usd'** might seem difficult to understand, but think of it this way:
> If you could buy each car individual km registered in the odometer, the value listed there would indicate how many km you would be able to buy with 1 dollar.

And following this concept, if we take into consideration 2 well known facts:
- Cars with more km on the odometer tend to be worn out thus in worse condition
- The more worn out the car is, it's expected that it's price would go down

We can then assume that the higher **'km per usd'** is, better that brand would be. On the other hand, if the it's low it means your dollars are paying a high price regardless of km ran.

*Now let's see what is the most common model for each brand.*

In [89]:
brand_common_model = {}

for brand in brands:
    brand_rows = autos_final['brand'] == brand
    common_model = autos_final.loc[brand_rows, 'model'].describe().top
    if brand_counts[brand] > 500:
        brand_common_model[brand] = common_model
    else:
        brand_others.append(brand)

brand_common_model['others'] = autos_final.loc[others_bool , 'model'].describe().top

In [90]:
brand_common_model

{'peugeot': '2_reihe',
 'bmw': '3er',
 'volkswagen': 'golf',
 'smart': 'fortwo',
 'ford': 'focus',
 'seat': 'ibiza',
 'renault': 'twingo',
 'mercedes_benz': 'c_klasse',
 'audi': 'a4',
 'opel': 'corsa',
 'mazda': '3_reihe',
 'toyota': 'yaris',
 'nissan': 'micra',
 'fiat': 'punto',
 'skoda': 'octavia',
 'citroen': 'andere',
 'others': 'andere'}

In [94]:
df_model = pd.DataFrame.from_dict(data=brand_common_model, orient='index', columns=['Common Model'])
df_brands = pd.concat([df_price_odo, df_model], axis=1)
df_brands

Mean Price  Mean Odometer  km per usd Common Model
peugeot        3100.692806  127122.302158   40.998032      2_reihe
bmw            8375.407663  132655.057719   15.838639          3er
volkswagen     5423.855787  128759.796438   23.739532         golf
smart          3591.021244   99590.288316   27.733138       fortwo
ford           3764.375687  124272.978789   33.012905        focus
seat           4423.116745  121303.066038   27.424794        ibiza
renault        2483.889193  128144.094847   51.590101       twingo
mercedes_benz  8683.419750  130927.329469   15.077853     c_klasse
audi           9373.771429  129254.658385   13.788971           a4
opel           2997.959438  129444.667201   43.177591        corsa
mazda          4112.596615  124464.033850   30.264100      3_reihe
toyota         5167.091062  115944.350759   22.438999        yaris
nissan         4750.063202  118370.786517   24.919834        micra
fiat           2818.456067  117066.945607   41.535842        punto
skoda          6393.018349  110884.665793   17.344650      octavia
citroen        3820.914713  119525.993884   31.282037       andere
others         8514.641212  112474.906514   13.209588       andere

**Takeaways**:

- Citroen and Others repeat the same common model, which raises a reason to go deeper and understand if further data cleaning is needed.